## Abhishek Sharma
### CS 2nd Year | Section : "I" | Roll no.: 01
### Enrollment No.: 12019009001127

## Advanced Artificial Intelligence and Machine Learning Lab 5

#### Faculty : PKB, MB, BM | Date : 23.03.2021

### Decision Tree for Regression :

Importing the Regressor Decision Tree Model : `from sklearn.tree import DecisionTreeRegressor`  

In [1]:
from sklearn.datasets import load_boston

In [2]:
import numpy as np
import pandas as pd

In [3]:
boston = load_boston()

In [4]:
data = pd.DataFrame(boston.data, columns = boston.feature_names)

In [5]:
data['MEDV'] = pd.DataFrame(boston.target) 

 **Working with two attributes of the `boston dataset`, they are `RM` and `ZN`**

In [6]:
x = data[['RM','ZN']]

In [7]:
y = data['MEDV']

In [8]:
from sklearn.model_selection import train_test_split

In [25]:
from sklearn.tree import DecisionTreeRegressor

In [11]:
dt = DecisionTreeRegressor()

In [12]:
x_train1, x_test1, y_train1, y_test1 = train_test_split(x,y, test_size = 0.2)

In [13]:
dt.fit(x_train1, y_train1)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=None, splitter='best')

In [15]:
yTestPredict = dt.predict(x_test1)

In [16]:
from sklearn.metrics import mean_squared_error

In [17]:
np.sqrt(mean_squared_error(y_test1, yTestPredict))

7.542256774659821

**Working with three attributes of `boston dataset`, they are `RM`, `ZN` and `LSTAT`** 

In [18]:
x = data[['RM','ZN', 'LSTAT']]

In [19]:
y = data['MEDV']

In [20]:
dt = DecisionTreeRegressor()

In [21]:
x_train1, x_test1, y_train1, y_test1 = train_test_split(x,y, test_size = 0.2)

In [22]:
dt.fit(x_train1, y_train1)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=None, splitter='best')

In [23]:
yTestPredict = dt.predict(x_test1)

In [24]:
np.sqrt(mean_squared_error(y_test1, yTestPredict))

5.293355064903285

### Two attribute model V/S Three attribute model using the Decision Tree regressor :

| Model Type | Mean Squared Error Value |
|:---:|:---:|
| Two Attribute Model | 7.54 |
| Three Attribute Model | 5.29|

### Decision Tree Classifier Model :

Importing the Decision Tree Classifier Model from the sklearn package : `from sklearn.tree import DecisionTreeClassifier`

In [34]:
from sklearn.tree import DecisionTreeClassifier

In [35]:
from sklearn.datasets import load_digits

In [36]:
import matplotlib.pyplot as plt

In [37]:
digits = load_digits()

In [38]:
x_train, x_test, y_train, y_test = train_test_split(digits.data,digits.target, test_size = 0.2)

In [39]:
dtc = DecisionTreeClassifier()

In [40]:
dtc.fit(x_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [41]:
dtc.score(x_test, y_test)

0.8527777777777777

Here it means that **85.28%** values of the ```test dataset``` matched with the ```train dataset``` during the fitting operation with 80-20 dataset division

In [42]:
x_train, x_test, y_train, y_test = train_test_split(digits.data,digits.target, test_size = 0.1)

In [43]:
dtc.fit(x_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [44]:
dtc.score(x_test, y_test)

0.8944444444444445

Here it means that **89.44%** values of the ```test dataset``` matched with the ```train dataset``` during the fitting operation with 90-10 dataset division

### Decision Tree Classifier Model Score with various division of data :

| Model Type | Model Score |
|:---:|:---:|
| 80 - 20 Division | 85.28% |
| 90 - 10 Division | 89.44% |

In [45]:
dtce = DecisionTreeClassifier(criterion = 'entropy')

In [46]:
dtce.fit(x_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='entropy',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [47]:
dtce.score(x_test, y_test)

0.9111111111111111

In [48]:
x_train, x_test, y_train, y_test = train_test_split(digits.data,digits.target, test_size = 0.2)

In [49]:
dtce.fit(x_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='entropy',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [50]:
dtce.score(x_test, y_test)

0.8527777777777777

In [51]:
x_train, x_test, y_train, y_test = train_test_split(digits.data,digits.target, test_size = 0.3)

In [52]:
dtce.fit(x_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='entropy',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [53]:
dtce.score(x_test, y_test)

0.85

### Decision Tree Classifier Model Score (criterion = entropy) with various division of data :

| Model Type | Model Score |
|:---:|:---:|
| 70 - 30 Division | 85.00% |
| 80 - 20 Division | 85.28% |
| 90 - 10 Division | 91.11% |

*******************************************
### <div align = 'center'> Thank You! </div>
